plutus-ir/src/Language/PlutusIR/Transform/LetFloat.hs

```diff
@@ -0,0 +1,455 @@
+
+Our algorithm is comprised of two passes of the PlutusIR AST:
+
+- 1st pass: Compute the "maximum ranks" of all lets in a given term. A maximum rank signifies a source-code position that the let will be moved/floated to during the 2nd pass.
+- 2nd pass: Move lets to their maximum rank positions. The pass makes sure that same-rank lets are grouped in multi-let-groups which
+are topologically sorted by their dependency graph (see 'Dependencies.runTermDeps').
+
+-}
+
+{- Note: [Float first-pass]
+
+Visits a PIR term and marks all identifiers introduced by lets with their "maximum rank" (level number in the paper).
+
+A maximum rank of a let-bound identifier can be seen as its "deepest" (lambda,Lamdba,or let)-dependency that occurs free in that let's RHS.
+
+The depth of a lambda/Lambda-dependency that occurs free in the RHS, is calculated as: the number of lambdas that surround this lambda/Lambda + 1.
+The depth of a let-dependency that occurs free in the RHS, is the depth of the "maximum rank" calculated for that let-dependency.
+
+Any dependencies that are bound inside the rhs are simply ignored.
+
+In the special case of a letrec group, every let of that group is assigned the maximum (deepest) rank among all lets.
+
+The first-pass returns a mapping of maximum-ranks to let-identifiers, sorted on depth (see 'FloatData').
+
+-}
+
+{- Note: [Float second-pass]
+
+The second-pass first cleans the input ASTs of all let-rhs'es and puts them in a separate table 'RhsTable'.
+(actually this is yet another pass, but we don't count this cleaning pass as an individual pass, to match the pass-numbering of the original algorithm).
+
+The second pass then traverses the 'cleaned term', and whenever it reaches a lambda/Lambda position (Rank) which exists as key in the 'FloatData' table,
+it floats its corresponding lets directly inside the lambda/Lambda's body.
+The introduced lets of each Rank (level number) are not grouped in a single multi-let-rec group (as in the original algorithm), but instead in possible-multiple
+let-(nonrec & rec) groups. This groupping is done in accordance with the dependency-graph (see 'runTermDeps'):
+any let group that depends on a prior let-(rec or nonrec) group, will appear nested inside that group. The algorithm merges adjacent let-nonrec groups
+into a single let-nonrec group, since the letnonrec is linearly scoped: `(let {i=...j=...} in...) === (let i=... in let j=... in ...)`.
+
+By design choice, the second-pass preserves the floated let-bindings' original (Rec and NonRec) recursivity labels;
+it will never demote a 'Rec' let-binding to a 'NonRec', despite the demotion being valid.
+The effect is that more let-groups may be generated in a specific Rank than it would otherwise be required (optimally).
+The demotion can potentially be a nice optimization/transformation, but the way the depenndency-graph is generated for datatype binds, makes this complicated.
```

Comment:
What exactly is complicated?

Reply:
Maybe complicated was a wrong choice of word. I rephrased it to "dependency graph does not currently support it"